src/state.py

```diff
@@ -0,0 +1,66 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for handling charm state."""
+
+import json
+
+
+class State:
+    """A magic state that uses a relation as the data store.
+
+    The get_relation callable is used to retrieve the relation.
+    As relation data values must be strings, all values are JSON encoded.
+    """
+
+    def __init__(self, app, get_relation):
+        """Construct.
+
+        Args:
+            app: workload application
+            get_relation: get peer relation method
+        """
+        # Use __dict__ to avoid calling __setattr__ and subsequent infinite recursion.
+        self.__dict__["_app"] = app
+        self.__dict__["_get_relation"] = get_relation
+
+    def __setattr__(self, name, value):
```

Comment:
This is actually being used across the charm for storing relation information (e.g. database and object storage credentials). An example of this can be seen under `relations/postgresql.py`.

Reply:
Thank you, found them. I'll leave comments there.